internal/controller/kubeconfig_provider_test.go

```diff
@@ -0,0 +1,129 @@
+
+	"github.com/kuadrant/dns-operator/api/v1alpha1"
+)
+
+var _ = Describe("Kubeconfig Provider", Labels{"multicluster"}, func() {
+
+	var (
+		secondaryClusterSecret *v1.Secret
+
+		// buffer containing all the log entries added during the current spec execution
+		// can be used in conjunction with `gbytes.Say` matcher to check log entries see https://onsi.github.io/gomega/#codegbytescode-testing-streaming-buffers
+		logBuffer *gbytes.Buffer
+	)
+
+	BeforeEach(func() {
+		logBuffer = gbytes.NewBuffer()
+		GinkgoWriter.TeeTo(logBuffer)
+
+		secondaryClusterSecret = &v1.Secret{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      "secondary-cluster-1",
+				Namespace: testDefaultClusterSecretNamespace,
+				Labels: map[string]string{
+					testDefaultClusterSecretLabel: "true",
+				},
+			},
+			StringData: map[string]string{
+				"kubeconfig": string(secondaryKubeconfig),
+			},
+		}
+	})
+
+	AfterEach(func() {
+		if secondaryClusterSecret != nil {
+			Expect(client.IgnoreNotFound(primaryK8sClient.Delete(ctx, secondaryClusterSecret))).To(Succeed())
+		}
+		GinkgoWriter.ClearTeeWriters()
+	})
+
+	It("successfully creates a new cluster from valid kubeconfig secret", Labels{"primary"}, func(ctx SpecContext) {
+		createClusterKubeconfigSecret(primaryK8sClient, secondaryClusterSecret, logBuffer)
+	})
+
+	It("successfully deletes cluster", Labels{"primary"}, func(ctx SpecContext) {
+		createClusterKubeconfigSecret(primaryK8sClient, secondaryClusterSecret, logBuffer)
+		deleteClusterKubeconfigSecret(primaryK8sClient, secondaryClusterSecret, logBuffer)
+	})
+
+	It("fails to create a new cluster from invalid kubeconfig secret", Labels{"primary"}, func(ctx SpecContext) {
+		invalidSecret := &v1.Secret{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      "secondary-cluster-1",
+				Namespace: testDefaultClusterSecretNamespace,
+				Labels: map[string]string{
+					testDefaultClusterSecretLabel: "true",
+				},
+			},
+			StringData: map[string]string{
+				"kubeconfig": "",
+			},
+		}
+		Expect(primaryK8sClient.Create(ctx, invalidSecret)).To(Succeed())
+		Eventually(logBuffer).Should(gbytes.Say(fmt.Sprintf("Secret does not contain kubeconfig data, skipping\\s*{\"cluster\": \"%s\", \"secret\": \"%s\\/%s\", \"key\": \"kubeconfig\"}", invalidSecret.Name, invalidSecret.Namespace, invalidSecret.Name)))
+	})
+
+	It("triggers reconcile of secondary cluster record on primary", Labels{"primary", "secondary"}, func(ctx SpecContext) {
+		createClusterKubeconfigSecret(primaryK8sClient, secondaryClusterSecret, logBuffer)
+
+		testNamespace := generateTestNamespaceName()
+		CreateNamespace(testNamespace, secondaryK8sClient)
+
+		testZoneDomainName := strings.Join([]string{GenerateName(), "example.com"}, ".")
+		testHostname := strings.Join([]string{"foo", testZoneDomainName}, ".")
+
+		secondaryRecord := &v1alpha1.DNSRecord{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      testHostname,
+				Namespace: testNamespace,
+			},
+			Spec: v1alpha1.DNSRecordSpec{
```

Comment:
Should this have "Delegate: true"? I guess it still emits the log either way, so that's why this is passing?

Reply:
> I guess it still emits the log either way, so that's why this is passing?

yes